File: java/v1/src/main/java/com/api/v1/services/customer/CustomerRegistrationServiceImpl.java
package com.api.v1.services.customer;

import com.api.v1.mappers.customer.CustomerResponseMapper;
import com.api.v1.builders.customer.CustomerBuilder;
import com.api.v1.domain.customer.Customer;
import com.api.v1.domain.customer.CustomerRepository;
import com.api.v1.dtos.customer.CustomerRegistrationRequestDto;
import com.api.v1.dtos.customer.CustomerResponseDto;
import com.api.v1.exceptions.customer.DuplicatedSsnException;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class CustomerRegistrationServiceImpl implements CustomerRegistrationService {

    @Autowired
    private CustomerRepository customerRepository;

    @Override
    public Mono<CustomerResponseDto> register(@Valid CustomerRegistrationRequestDto request) {
        return customerRepository
                .findAll()
                .filter(e -> e.getSsn().equals(request.ssn()))
                .hasElements()
                .flatMap(exists -> {
                    if (exists) return Mono.error(new DuplicatedSsnException(request.ssn()));
                    return Mono.defer(() -> {
                        Customer customer = CustomerBuilder.create().fromDto(request).build();
                        return customerRepository.save(customer);
                    })
                    .flatMap(e -> Mono.just(new CustomerResponseMapper(e).mapToDto()));
                });
    }

}

File: v2/src/main/kotlin/com/api/v2/customer/services/CustomerRegistrationServiceImpl.kt
package com.api.v2.customer.services

import com.api.v2.customer.builders.CustomerBuilder
import com.api.v2.customer.domain.CustomerRepository
import com.api.v2.customer.dtos.CustomerRegistrationRequestDto
import com.api.v2.customer.dtos.CustomerResponseDto
import com.api.v2.customer.exceptions.DuplicatedSsnException
import com.api.v2.customer.mappers.CustomerResponseMapper
import jakarta.validation.Valid
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class CustomerRegistrationServiceImpl: CustomerRegistrationService {

    @Autowired
    private lateinit var customerRepository: CustomerRepository

    override suspend fun register(requestDto: @Valid CustomerRegistrationRequestDto): CustomerResponseDto {
        return withContext(Dispatchers.IO) {
            if (checkTheExistenceOfDuplicatedSsn(requestDto.ssn)) {
                throw DuplicatedSsnException(requestDto.ssn)
            }
            val newCustomer = CustomerBuilder.create().fromDto(requestDto).build()
            val savedCustomer = customerRepository.save(newCustomer)
            CustomerResponseMapper.mapToDto(savedCustomer)
        }
    }

    private suspend fun checkTheExistenceOfDuplicatedSsn(ssn: String): Boolean {
        return customerRepository.findAll().firstOrNull { e -> e.ssn == ssn } != null
    }

}